pkg/providers/vsphere/vmlifecycle/update_status.go

```diff
@@ -1897,5 +1853,238 @@ func reconcileStatusController(
+		vmCtx.VM,
+		vmopv1.VirtualMachineHardwareDeviceConfigVerified,
+		vmopv1.VirtualMachineHardwareDeviceConfigMismatchReason,
+		"%s",
+		apierrorsutil.NewAggregate(errs).Error())
+}
+
+// checkControllers verifies that all controllers specified in the VM spec are
+// attached and no extra controllers exist.
+func checkControllers(vm *vmopv1.VirtualMachine) []error {
+	var (
+		errs                []error
+		expectedControllers = sets.New[pkgutil.ControllerID]()
+		actualControllers   = sets.New[pkgutil.ControllerID]()
+	)
+
+	// Build expected controllers from spec.
+	if hw := vm.Spec.Hardware; hw != nil {
+		for _, ctrl := range hw.IDEControllers {
+			expectedControllers.Insert(pkgutil.ControllerID{
+				ControllerType: vmopv1.VirtualControllerTypeIDE,
+				BusNumber:      ctrl.BusNumber,
+			})
+		}
+		for _, ctrl := range hw.NVMEControllers {
+			expectedControllers.Insert(pkgutil.ControllerID{
+				ControllerType: vmopv1.VirtualControllerTypeNVME,
+				BusNumber:      ctrl.BusNumber,
+			})
+		}
+		for _, ctrl := range hw.SATAControllers {
+			expectedControllers.Insert(pkgutil.ControllerID{
+				ControllerType: vmopv1.VirtualControllerTypeSATA,
+				BusNumber:      ctrl.BusNumber,
+			})
+		}
+		for _, ctrl := range hw.SCSIControllers {
+			expectedControllers.Insert(pkgutil.ControllerID{
+				ControllerType: vmopv1.VirtualControllerTypeSCSI,
+				BusNumber:      ctrl.BusNumber,
+			})
+		}
+	}
+
+	// Build actual controllers from status.
+	if vm.Status.Hardware != nil {
+		for _, ctrl := range vm.Status.Hardware.Controllers {
+			actualControllers.Insert(pkgutil.ControllerID{
+				ControllerType: ctrl.Type,
+				BusNumber:      ctrl.BusNumber,
+			})
+		}
+	}
+
+	errs = append(errs, pkgutil.CompareAndReportMissingExtra(
+		expectedControllers, actualControllers,
+		"missing controllers", "unexpected controllers")...)
+
+	return errs
+}
+
+// checkPVCVolumes verifies that all PVC volumes in the spec are attached and
+// no extra disks are attached that aren't in the spec.
+func checkPVCVolumes(vm *vmopv1.VirtualMachine) []error {
+
+	var (
+		errs            []error
+		expectedVolumes = sets.New[pkgutil.DevicePlacement]()
+		actualVolumes   = sets.New[pkgutil.DevicePlacement]()
+	)
+
+	for _, vol := range vm.Spec.Volumes {
+		if vol.PersistentVolumeClaim == nil ||
+			vol.PersistentVolumeClaim.UnmanagedVolumeClaim != nil {
+			continue
+		}
+
+		pvcSpec := vol.PersistentVolumeClaim
+		if pvcSpec.ControllerType == "" ||
+			pvcSpec.ControllerBusNumber == nil ||
+			pvcSpec.UnitNumber == nil {
+			errs = append(errs, fmt.Errorf(
+				"PVC volume %q has incomplete placement information",
+				vol.Name))
+			continue
+		}
+
+		expectedVolumes.Insert(pkgutil.DevicePlacement{
+			Key:                 vol.Name,
+			ControllerType:      pvcSpec.ControllerType,
+			ControllerBusNumber: *pvcSpec.ControllerBusNumber,
+			UnitNumber:          *pvcSpec.UnitNumber,
+		})
+	}
+
+	// Build mapping from disk UUID to volume name for attached PVC volumes.
```

Comment:
Hi @akutz , @bryanv , to reduce the duplicate code between the batch volume controller and here for the disk UUID to volume name lookup, I am thinking about using the volume status directly here. 

This introduces a race condition, but since we eventually want to move the v2 volume controller to the virtual machine controller anyway. 

Do you folks think it is acceptable? 

If not, I could fall back to the CNSBatchAttachmentName CR look-up.